src/network/netdev/bond.c

```diff
@@ -1,16 +1,12 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
-#include <netinet/ether.h>
-#include <linux/if_bonding.h>
```

Comment:
Unfortunately this doesn't build on F30 with kernel-headers-5.0.6-300.fc30.x86_64. Again, glibc and kernel cannot figure out how to make compatible headers.

```
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/linux/in.h:29:3: error: redeclaration of enumerator ‘IPPROTO_IP’
   29 |   IPPROTO_IP = 0,  /* Dummy protocol for TCP  */
      |   ^~~~~~~~~~
/usr/include/netinet/in.h:42:5: note: previous definition of ‘IPPROTO_IP’ was here
   42 |     IPPROTO_IP = 0,    /* Dummy protocol for TCP.  */
      |     ^~~~~~~~~~
/usr/include/linux/in.h:31:3: error: redeclaration of enumerator ‘IPPROTO_ICMP’
   31 |   IPPROTO_ICMP = 1,  /* Internet Control Message Protocol */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:44:5: note: previous definition of ‘IPPROTO_ICMP’ was here
   44 |     IPPROTO_ICMP = 1,    /* Internet Control Message Protocol.  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:33:3: error: redeclaration of enumerator ‘IPPROTO_IGMP’
   33 |   IPPROTO_IGMP = 2,  /* Internet Group Management Protocol */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:46:5: note: previous definition of ‘IPPROTO_IGMP’ was here
   46 |     IPPROTO_IGMP = 2,    /* Internet Group Management Protocol. */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:35:3: error: redeclaration of enumerator ‘IPPROTO_IPIP’
   35 |   IPPROTO_IPIP = 4,  /* IPIP tunnels (older KA9Q tunnels use 94) */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:48:5: note: previous definition of ‘IPPROTO_IPIP’ was here
   48 |     IPPROTO_IPIP = 4,    /* IPIP tunnels (older KA9Q tunnels use 94).  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:37:3: error: redeclaration of enumerator ‘IPPROTO_TCP’
   37 |   IPPROTO_TCP = 6,  /* Transmission Control Protocol */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:50:5: note: previous definition of ‘IPPROTO_TCP’ was here
   50 |     IPPROTO_TCP = 6,    /* Transmission Control Protocol.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:39:3: error: redeclaration of enumerator ‘IPPROTO_EGP’
   39 |   IPPROTO_EGP = 8,  /* Exterior Gateway Protocol  */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:52:5: note: previous definition of ‘IPPROTO_EGP’ was here
   52 |     IPPROTO_EGP = 8,    /* Exterior Gateway Protocol.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:41:3: error: redeclaration of enumerator ‘IPPROTO_PUP’
   41 |   IPPROTO_PUP = 12,  /* PUP protocol    */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:54:5: note: previous definition of ‘IPPROTO_PUP’ was here
   54 |     IPPROTO_PUP = 12,    /* PUP protocol.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:43:3: error: redeclaration of enumerator ‘IPPROTO_UDP’
   43 |   IPPROTO_UDP = 17,  /* User Datagram Protocol  */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:56:5: note: previous definition of ‘IPPROTO_UDP’ was here
   56 |     IPPROTO_UDP = 17,    /* User Datagram Protocol.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:45:3: error: redeclaration of enumerator ‘IPPROTO_IDP’
   45 |   IPPROTO_IDP = 22,  /* XNS IDP protocol   */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:58:5: note: previous definition of ‘IPPROTO_IDP’ was here
   58 |     IPPROTO_IDP = 22,    /* XNS IDP protocol.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:47:3: error: redeclaration of enumerator ‘IPPROTO_TP’
   47 |   IPPROTO_TP = 29,  /* SO Transport Protocol Class 4 */
      |   ^~~~~~~~~~
/usr/include/netinet/in.h:60:5: note: previous definition of ‘IPPROTO_TP’ was here
   60 |     IPPROTO_TP = 29,    /* SO Transport Protocol Class 4.  */
      |     ^~~~~~~~~~
/usr/include/linux/in.h:49:3: error: redeclaration of enumerator ‘IPPROTO_DCCP’
   49 |   IPPROTO_DCCP = 33,  /* Datagram Congestion Control Protocol */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:62:5: note: previous definition of ‘IPPROTO_DCCP’ was here
   62 |     IPPROTO_DCCP = 33,    /* Datagram Congestion Control Protocol.  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:51:3: error: redeclaration of enumerator ‘IPPROTO_IPV6’
   51 |   IPPROTO_IPV6 = 41,  /* IPv6-in-IPv4 tunnelling  */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:64:5: note: previous definition of ‘IPPROTO_IPV6’ was here
   64 |     IPPROTO_IPV6 = 41,     /* IPv6 header.  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:53:3: error: redeclaration of enumerator ‘IPPROTO_RSVP’
   53 |   IPPROTO_RSVP = 46,  /* RSVP Protocol   */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:66:5: note: previous definition of ‘IPPROTO_RSVP’ was here
   66 |     IPPROTO_RSVP = 46,    /* Reservation Protocol.  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:55:3: error: redeclaration of enumerator ‘IPPROTO_GRE’
   55 |   IPPROTO_GRE = 47,  /* Cisco GRE tunnels (rfc 1701,1702) */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:68:5: note: previous definition of ‘IPPROTO_GRE’ was here
   68 |     IPPROTO_GRE = 47,    /* General Routing Encapsulation.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:57:3: error: redeclaration of enumerator ‘IPPROTO_ESP’
   57 |   IPPROTO_ESP = 50,  /* Encapsulation Security Payload protocol */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:70:5: note: previous definition of ‘IPPROTO_ESP’ was here
   70 |     IPPROTO_ESP = 50,      /* encapsulating security payload.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:59:3: error: redeclaration of enumerator ‘IPPROTO_AH’
   59 |   IPPROTO_AH = 51,  /* Authentication Header protocol */
      |   ^~~~~~~~~~
/usr/include/netinet/in.h:72:5: note: previous definition of ‘IPPROTO_AH’ was here
   72 |     IPPROTO_AH = 51,       /* authentication header.  */
      |     ^~~~~~~~~~
/usr/include/linux/in.h:61:3: error: redeclaration of enumerator ‘IPPROTO_MTP’
   61 |   IPPROTO_MTP = 92,  /* Multicast Transport Protocol  */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:74:5: note: previous definition of ‘IPPROTO_MTP’ was here
   74 |     IPPROTO_MTP = 92,    /* Multicast Transport Protocol.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:63:3: error: redeclaration of enumerator ‘IPPROTO_BEETPH’
   63 |   IPPROTO_BEETPH = 94,  /* IP option pseudo header for BEET */
      |   ^~~~~~~~~~~~~~
/usr/include/netinet/in.h:76:5: note: previous definition of ‘IPPROTO_BEETPH’ was here
   76 |     IPPROTO_BEETPH = 94,   /* IP option pseudo header for BEET.  */
      |     ^~~~~~~~~~~~~~
/usr/include/linux/in.h:65:3: error: redeclaration of enumerator ‘IPPROTO_ENCAP’
   65 |   IPPROTO_ENCAP = 98,  /* Encapsulation Header   */
      |   ^~~~~~~~~~~~~
/usr/include/netinet/in.h:78:5: note: previous definition of ‘IPPROTO_ENCAP’ was here
   78 |     IPPROTO_ENCAP = 98,    /* Encapsulation Header.  */
      |     ^~~~~~~~~~~~~
/usr/include/linux/in.h:67:3: error: redeclaration of enumerator ‘IPPROTO_PIM’
   67 |   IPPROTO_PIM = 103,  /* Protocol Independent Multicast */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:80:5: note: previous definition of ‘IPPROTO_PIM’ was here
   80 |     IPPROTO_PIM = 103,    /* Protocol Independent Multicast.  */
      |     ^~~~~~~~~~~
/usr/include/linux/in.h:69:3: error: redeclaration of enumerator ‘IPPROTO_COMP’
   69 |   IPPROTO_COMP = 108,  /* Compression Header Protocol  */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:82:5: note: previous definition of ‘IPPROTO_COMP’ was here
   82 |     IPPROTO_COMP = 108,    /* Compression Header Protocol.  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:71:3: error: redeclaration of enumerator ‘IPPROTO_SCTP’
   71 |   IPPROTO_SCTP = 132,  /* Stream Control Transport Protocol */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:84:5: note: previous definition of ‘IPPROTO_SCTP’ was here
   84 |     IPPROTO_SCTP = 132,    /* Stream Control Transmission Protocol.  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:73:3: error: redeclaration of enumerator ‘IPPROTO_UDPLITE’
   73 |   IPPROTO_UDPLITE = 136, /* UDP-Lite (RFC 3828)   */
      |   ^~~~~~~~~~~~~~~
/usr/include/netinet/in.h:86:5: note: previous definition of ‘IPPROTO_UDPLITE’ was here
   86 |     IPPROTO_UDPLITE = 136, /* UDP-Lite protocol.  */
      |     ^~~~~~~~~~~~~~~
/usr/include/linux/in.h:75:3: error: redeclaration of enumerator ‘IPPROTO_MPLS’
   75 |   IPPROTO_MPLS = 137,  /* MPLS in IP (RFC 4023)  */
      |   ^~~~~~~~~~~~
/usr/include/netinet/in.h:88:5: note: previous definition of ‘IPPROTO_MPLS’ was here
   88 |     IPPROTO_MPLS = 137,    /* MPLS in IP.  */
      |     ^~~~~~~~~~~~
/usr/include/linux/in.h:77:3: error: redeclaration of enumerator ‘IPPROTO_RAW’
   77 |   IPPROTO_RAW = 255,  /* Raw IP packets   */
      |   ^~~~~~~~~~~
/usr/include/netinet/in.h:90:5: note: previous definition of ‘IPPROTO_RAW’ was here
   90 |     IPPROTO_RAW = 255,    /* Raw IP packets.  */
      |     ^~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:79:3: error: redeclaration of enumerator ‘IPPROTO_MAX’
   79 |   IPPROTO_MAX
      |   ^~~~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:92:5: note: previous definition of ‘IPPROTO_MAX’ was here
   92 |     IPPROTO_MAX
      |     ^~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:85:8: error: redefinition of ‘struct in_addr’
   85 | struct in_addr {
      |        ^~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:31:8: note: originally defined here
   31 | struct in_addr
      |        ^~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:168:8: error: redefinition of ‘struct ip_mreq’
  168 | struct ip_mreq  {
      |        ^~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:264:8: note: originally defined here
  264 | struct ip_mreq
      |        ^~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:173:8: error: redefinition of ‘struct ip_mreqn’
  173 | struct ip_mreqn {
      |        ^~~~~~~~
In file included from /usr/include/netinet/in.h:37,
                 from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/bits/in.h:149:8: note: originally defined here
  149 | struct ip_mreqn
      |        ^~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:179:8: error: redefinition of ‘struct ip_mreq_source’
  179 | struct ip_mreq_source {
      |        ^~~~~~~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:273:8: note: originally defined here
  273 | struct ip_mreq_source
      |        ^~~~~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:185:8: error: redefinition of ‘struct ip_msfilter’
  185 | struct ip_msfilter {
      |        ^~~~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:323:8: note: originally defined here
  323 | struct ip_msfilter
      |        ^~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:197:8: error: redefinition of ‘struct group_req’
  197 | struct group_req {
      |        ^~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:300:8: note: originally defined here
  300 | struct group_req
      |        ^~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:202:8: error: redefinition of ‘struct group_source_req’
  202 | struct group_source_req {
      |        ^~~~~~~~~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:309:8: note: originally defined here
  309 | struct group_source_req
      |        ^~~~~~~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:208:8: error: redefinition of ‘struct group_filter’
  208 | struct group_filter {
      |        ^~~~~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:344:8: note: originally defined here
  344 | struct group_filter
      |        ^~~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:222:8: error: redefinition of ‘struct in_pktinfo’
  222 | struct in_pktinfo {
      |        ^~~~~~~~~~
In file included from /usr/include/netinet/in.h:37,
                 from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/bits/in.h:157:8: note: originally defined here
  157 | struct in_pktinfo
      |        ^~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:13,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in.h:232:8: error: redefinition of ‘struct sockaddr_in’
  232 | struct sockaddr_in {
      |        ^~~~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:237:8: note: originally defined here
  237 | struct sockaddr_in
      |        ^~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:14,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in6.h:33:8: error: redefinition of ‘struct in6_addr’
   33 | struct in6_addr {
      |        ^~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:211:8: note: originally defined here
  211 | struct in6_addr
      |        ^~~~~~~~
In file included from /usr/include/linux/l2tp.h:14,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in6.h:50:8: error: redefinition of ‘struct sockaddr_in6’
   50 | struct sockaddr_in6 {
      |        ^~~~~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:252:8: note: originally defined here
  252 | struct sockaddr_in6
      |        ^~~~~~~~~~~~
In file included from /usr/include/linux/l2tp.h:14,
                 from ../src/network/netdev/l2tp-tunnel.h:4,
                 from ../src/network/netdev/netdev.c:16:
/usr/include/linux/in6.h:60:8: error: redefinition of ‘struct ipv6_mreq’
   60 | struct ipv6_mreq {
      |        ^~~~~~~~~
In file included from ../src/basic/in-addr-util.h:4,
                 from ../src/network/netdev/bond.h:6,
                 from ../src/network/netdev/netdev.c:10:
/usr/include/netinet/in.h:288:8: note: originally defined here
  288 | struct ipv6_mreq
      |        ^~~~~~~~~
```

This needs some work-around.

Reply:
Sorry, I tested with the kernel headers provided by centos 7. I hope now the builds succeed.